service/src/LevelZero.cpp

```diff
@@ -523,6 +526,366 @@ namespace geopm
+                        __LINE__);
+
+        std::vector<uint32_t> metric_count(data_count);
+        std::vector<zet_typed_value_t> metric_values(num_metric_values);
+        ze_result = zetMetricGroupCalculateMultipleMetricValuesExp(m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx), calculation_type, data_size, data.data(), &data_count, &num_metric_values, metric_count.data(), metric_values.data());
+//        check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+//                        "LevelZero::" + std::string(__func__) +
+//                        ": LevelZero Metric group calculate metric values to calculate data failed",
+//                        __LINE__);
+//        //std::cout << "\tmetric group calculate metric values: " << std::to_string(num_metric_values) << std::endl;
+//        ze_result = ZE_RESULT_ERROR_UNKNOWN;
+
+        uint32_t num_metric = m_devices.at(l0_device_idx).subdevice.num_metric.at(l0_domain_idx);
+        if (ze_result == ZE_RESULT_SUCCESS) {
+            unsigned int num_reports = num_metric_values / num_metric;
+
+            for (unsigned int metric_idx = 0; metric_idx < num_metric; metric_idx++)
+            {
+                //TODO: It is possible that simply parsing all the metrics is
+                //      faster than the additional API calls to check the metric
+                //      name.  This should be studied
+                std::vector<zet_metric_handle_t> metric_handle(num_metric);
+                ze_result = zetMetricGet(m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx),
+                                         &num_metric, metric_handle.data());
+                check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+                                "LevelZero::" + std::string(__func__) +
+                                ": LevelZero Metric handle acquisition failed",
+                                __LINE__);
+
+                // process metrics
+                zet_metric_properties_t metric_properties;
+                ze_result = zetMetricGetProperties(metric_handle.at(metric_idx), &metric_properties);
+                check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+                                "LevelZero::" + std::string(__func__) +
+                                ": LevelZero Metric property acquisition failed",
+                                __LINE__);
+
+                std::string metric_name (metric_properties.name);
+
+                //TODO: check timing impact of only processing metrics supported by IOGroup
+                if(metric_name.compare("XVE_ACTIVE") == 0 ||
+                   metric_name.compare("XVE_STALL") == 0) {
+                    for (unsigned int report_idx = 0; report_idx < num_reports; report_idx++)
+                    {
+                        //This is the actual gathering of the data
+                        zet_typed_value_t data = metric_values.at(report_idx*num_metric + metric_idx);
```

Comment:
Preferably, it would be better if a data structure was used that more naturally represented the data being stored. If it's not convenient for whatever reason, please add a comment explaining this formula and how `metric_values` is structured such that this formula correctly indexes into the data item of interest.

I'm a bit nervous that `report_idx` depends on `num_reports` which is `num_metric_values / num_metric`. Does this formula take into account what happens when `num_metric_values` is not divisible by `num_metric`? If needed, please add checks and/or assertions to ensure any preconditions needed for this formula to work.